mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml.feature
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.ml.util.{DefaultReadWriteTest}
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.mllib.util.TestingUtils._
+import org.apache.spark.sql.Row
+
+class ImputerSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultReadWriteTest {
+
+  test("Imputer for Double with default missing Value NaN") {
```

Comment:
The way these tests are written was pretty confusing to me. It seems like the "mean" column should be the mean of some _values_ but really it is the expected output when "mean" strategy is used. This is minor since it just affects readability and might be more of a personal preference. I think the following is clearer:

```scala
val df = sqlContext.createDataFrame( Seq(
      (0, 1.0),
      (1, 1.0),
      (2, 3.0),
      (3, 4.0),
      (4, Double.NaN)
    )).toDF("id", "value")
    val expectedOutput = Map(
      "mean"->Array(1.0, 1.0, 3.0, 4.0, 2.25),
      "median" -> Array(1.0, 1.0, 3.0, 4.0, 1.0),
      "mode" -> Array(1.0, 1.0, 3.0, 4.0, 1.0))
    Seq("mean", "median", "mode").foreach { strategy =>
      val imputer = new Imputer().setInputCol("value").setOutputCol("out").setStrategy(strategy)
      val model = imputer.fit(df)
      val result = model.transform(df).select("out").collect().map(_.getDouble(0))
      result.zip(expectedOutput(strategy)).foreach { case (actual, expected) =>
        assert(actual ~== expected absTol 1e-5)
      }
```
Really, just any way of indicating that the extra columns are _expected outputs_ would be clearer to me.